src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Healthchecks for sources
+use anyhow::Context;
+use chrono::NaiveDateTime;
+use std::fmt::Display;
+use std::sync::Arc;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+use tokio::sync::Mutex;
+use tracing::trace;
+
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::{Listen, ListenEvent, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_repr::{Datum, GlobalId, Row, Timestamp};
+
+use crate::source::{CollectionMetadata, NowFn};
+use crate::types::sources::SourceData;
+
+/// The Healthchecker is responsible for tracking the current state
+/// of a Timely worker for a source, as well as updating the relevant
+/// state collection based on it.
+pub struct Healthchecker {
+    /// Name of the source (e.g. kafka-s1)
+    source_name: String,
+    /// Name of the upstream resource, if any (e.g. Kafka topic, file path)
+    upstream_name: Option<String>,
+    /// Internal ID of the source (e.g. s1)
+    source_id: String,
+    /// Type of the source, from SourceConnection::name
+    source_type: &'static str,
+    /// The ID of the Timely worker on which this operator is executing
+    worker_id: usize,
+    /// The total count of Timely workers
+    worker_count: usize,
+    /// Whether this is an active Timely worker or not
+    active: bool,
+    /// Current status of this source
+    current_status: SourceStatus,
+    /// Last observed upper
+    upper: Antichain<Timestamp>,
+    /// Write handle of the Healthchecker persist shard
+    write_handle: WriteHandle<SourceData, (), Timestamp, i64>,
```

Comment:
This is `SourceData` because that's what's in all the other shards that are backing sources or tables, right?

Reply:
Exactly! Should I add a comment about it? This is how all our tables (and I believe sources as well) are persisted.

Reply:
probably would be good, yes